examples.tel

```diff
@@ -4,12 +4,12 @@
 main = \input -> ("Hello, World!", 0)
 
 -- Simple example of ad-hoc user defined types
--- MyInt = let intTag = unique
---         in ( \i -> if not i
---                    then "MyInt must not be 0"
---                    else (intTag, i)
---            , \i -> if dEqual (left i) intTag
---                    then 0
---                    else "expecting MyInt"
---            )
+-- MyInt = let wrapper = \h -> ( \i -> if not i
+--                               	    then "MyInt must n"
+--                                     else  i
+--                             , \i -> if dEqual (left i)
+--                                     then 0
+--                                     else "expecting MyInt"
+--                             )
+--         in wrapper (# wrapper)
```

Comment:
The `h` argument is intended to work like the `intTag` argument did in the old code. The idea is that these custom data types with smart constructors are identified with a "tag" which is simply an integer.

So in Haskell terms, custom data types are simply pairs (`(Int, a)`) where `a` is the "data" of the type, and `Int` is the hash to uniquely identify the type. This `Int` can then be checked statically so that it acts like type checking to the programmer.